### DISCUSSION TITLE
Question about Changing the Source / Outfitter Hotkey Changes

### DISCUSSION THREAD

**Discussion by FurretGoWalcc** [endless-sky/endless-sky#8761] [Balance, Mechanics, and Interactions]
Howdy everyone. 

I wanted to ask about a question about how to go about editing the source files. I am new to this community, and as far as I know, there isn't any guidelines or stylistic stuff about the source files that is needed to be followed, so wanted to clear my ideas before putting in the work and making a PR with some commits. 

In relation to #8735, there are 3 classes that I'm looking to change, ShopPanel, and it's two derived classes, OutfitterPanel, and ShipyardPanel. I'm not super solid on the Sprite and GUI system that ES uses, so while I'll try my best to get something functional, someone might have to come along later and build something a bit prettier/more friendly later on. 

Changes proposed:
1. Alter KeyDown in the base to call a new virtual function, TransactionHandle, when one of the purchase buttons is pressed (b,i,s,c,u) which will pass along the key that was pressed and return a BuyResult. Each of the derived classes will override and implement this function. This should allow me to leave the Shipyard Buy/Sell code as close to the current implementation possible while altering the OutfitterPanel code.
2. Remove several virtual protected functions, CanBuy, CanSell, Buy, Sell, which are implemented in the derived classes. These are being removed to provide more options on outfit sources and destinations (storage, installed, cargo, store), with much of the logic being translated into new functions. As stated previously, I will likely keep these 4 functions defined privately within the ShipyardPanel, and called in it's version of TransactionHandle
3. In OutfitterPanel, I plan on replacing the those 4 functions with several new ones. CanSource, verify outfit can be taken from wherever specified, returns a BuyResult, CanDestination, verify outfit can be installed/stored/cargo'd/sold, returns a BuyResult. Then once that is done transaction handle will complete the transaction, adjusting variables as needed, otherwise returning the BuyResults of the failed Can* functions back up to KeyDown, which will display it as it currently displays CanBuy.

Peace.

**Comment by bene-dictator** [comment]
I'm not involved with the code-side of Endless Sky, but a good place to start may be the endless-sky wiki. The guidelines and style requirements can be found there.

From what I know, GUI is hard to touch because Michael Zahnhiser, the creator of Endless Sky (who has stepped away from development for now), made almost all of the UI graphics and we have not been able to replicate it well enough. Other people may know a lot better about the situation – you can ask them on the community discord server, especially on `#content-creating` or `#the-hub`.

**Comment by warp-core** [comment]
The files in the [docs](https://github.com/endless-sky/endless-sky/tree/master/docs) folder might be useful to you in getting set up for editing the source code.

There is a [style guide](https://endless-sky.github.io/styleguide/styleguide.xml), linked near the bottom of the [wiki](https://github.com/endless-sky/endless-sky/wiki).

As for the proposed redesign, calling a new method when one of the "transaction" keys is pressed sounds good, as do the new methods you suggest for OurfitterPanel, though I may have more comments after actually seeing the implementation.
